embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/logs/LogOrchestrator.kt

```diff
@@ -0,0 +1,78 @@
+package io.embrace.android.embracesdk.internal.logs
+
+import androidx.annotation.VisibleForTesting
+import io.embrace.android.embracesdk.comms.delivery.DeliveryService
+import io.embrace.android.embracesdk.internal.clock.Clock
+import io.embrace.android.embracesdk.worker.ScheduledWorker
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+
+internal class LogOrchestrator(
+    private val logOrchestratorScheduledWorker: ScheduledWorker,
+    private val clock: Clock,
+    private val sink: LogSinkImpl,
+    private val deliveryService: DeliveryService
+) {
+    private var lastLogTime: Long = 0
+    private var inactivityFuture: ScheduledFuture<*>? = null
+    private var batchTimeFuture: ScheduledFuture<*>? = null
+
+    init {
+        sink.callOnLogsStored(::onLogsAdded)
+    }
+    private fun onLogsAdded() {
+        lastLogTime = clock.now()
+        if (sink.completedLogs().size >= MAX_LOGS_PER_BATCH) {
+            sendLogs()
+        } else {
+            scheduleInactivityCheck()
+            if (batchTimeFuture == null) {
+                scheduleBatchTimeCheck()
+            }
+        }
+    }
+
+    @VisibleForTesting
+    fun sendLogs() {
+        inactivityFuture?.cancel(false)
+        inactivityFuture = null
+        batchTimeFuture?.cancel(false)
+        batchTimeFuture = null
+
+        // Sink is synchronized, so even if sendLogs is called concurrently, the logs
+        // will be sent only once.
+        val storedLogs = sink.flushLogs()
+
+        if (storedLogs.isNotEmpty()) {
+            deliveryService.sendLogs(LogPayload(logs = storedLogs))
+        }
+    }
+
+    private fun scheduleInactivityCheck() {
```

Comment:
Do we need two jobs? If we only schedule the next time the batch should be checked based on the current state of the batch and the time it was first created, would that be sufficient?

So basically, every time we get a new log, we cancel the next check and schedule a new one that is the minimum of the inactivity time and the time since the first log in the batch was added.

And when the check runs, all it has to do is see if either the inactivity time has elapsed or if the total batch time has been exceed.

If the log results in the max batch size being reached, send the batch cancel the check, and let the next time a log comes in to schedule another one.

Reply:
Changed to be one job only.